File: AoC-2015/Day20/Problem1.py
with open("Day20/input.txt") as f:
    min_gifts = int(f.readline())

n = 1_000_000
s = [1]*(n+1)
s[0] = s[1] = 0

for i in range(2,int(n**0.5)+1):
    if s[i]:
        s[i*i:n+1:i] = [0]*((n-i*i)//i+1)
primes = [i for i in range(n+1) if s[i]]

gifts, house = 10, 1

while gifts < min_gifts:
    house += 1
    sigma, idx, num = 1, 0, house

    while idx < len(primes) and primes[idx] * primes[idx] <= num:
        count = 0
        while num % primes[idx] == 0:
            count += 1
            num //= primes[idx]

        if count > 0:
            sigma *= (primes[idx]**(count+1) - 1) // (primes[idx]-1)

        idx += 1

    if num > 1:
        sigma *= (num + 1)

    gifts = 10 * sigma

print(house)

File: AoC-2025/Day02/Problem1.java
import java.io.FileReader;
import java.io.IOException;
import java.io.BufferedReader;
import java.util.ArrayList;

class Range
{

    int minShift;
    String prefix, left, right;

    Range (StringBuilder sb)
    {
        String range = sb.toString();
        int idx = range.indexOf("-");
        
        this.left = range.substring(0,idx);
        this.right = range.substring(idx+1);
        
        int index = 0;
        while(index < left.length() && left.charAt(index) == right.charAt(index)) index++;

        this.prefix = left.substring(0,index);
        this.minShift = left.length() - prefix.length();
    }
}

public class Problem1
{
    public static void main(String[] args) {solve();}

    public static void solve()
    {
        try(BufferedReader br = new BufferedReader(new FileReader("Day02/input.txt")))
        {   
            String line = br.readLine();
            ArrayList<Range> ranges = parseLine(line);

            long ans = 0;

            for(Range rng: ranges){
                long l = Long.parseLong(rng.left);
                long r = Long.parseLong(rng.right);

                for(long n = l; n <= r; n++){
                    String suffix = Long.toString(n);
                    while (suffix.length() < rng.minShift) suffix = "0" + suffix;

                    long num = Long.parseLong(suffix);

                    if (invalid(num)) ans += num;
                }

            }

            System.out.println(ans);
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    public static boolean invalid(long num)
    {   
        
        String s = Long.toString(num);
        int n = s.length();

        if(n%2 == 1) return false;

        for(int i=0;i<n/2;i++){
            if(s.charAt(i) != s.charAt(i+n/2)) return false;
        }

        return true;
    }

    public static ArrayList<Range> parseLine(String line)
    {
        ArrayList<Range> arr = new ArrayList<>();
        StringBuilder range = new StringBuilder();
        int n = line.length(), prev = 0;

        for(int i=0;i<=n;i++){
            if(i == n || line.charAt(i) == ','){
                arr.add(new Range(range));
                range = new StringBuilder();
            }
            else range.append(line.charAt(i));
        }

        return arr;
    }
}